packages/zk-certificates/lib/keyManagement.ts

```diff
@@ -2,22 +2,51 @@
 import {
   eddsaKeyGenerationMessage,
   eddsaPrimeFieldMod,
+  EddsaPrivateKey,
 } from '@galactica-net/galactica-types';
 import createBlakeHash from 'blake-hash';
 import { Buffer } from 'buffer';
+import { Eddsa } from 'circomlibjs';
 import { Signer } from 'ethers';
 import { Scalar, utils } from 'ffjavascript';
 
 /**
  * Generates the eddsa private key from the ethereum private key signing a fixed message.
  *
  * @param signer - Ethers signer.
- * @returns The eddsa private key.
+ * @returns The eddsa private key as buffer.
  */
 export async function getEddsaKeyFromEthSigner(
   signer: Signer,
-): Promise<string> {
-  return signer.signMessage(eddsaKeyGenerationMessage);
+): Promise<EddsaPrivateKey> {
+  // use signature as entropy input so that the EdDSA key can be derived from the Ethereum private key
+  const signature = await signer.signMessage(eddsaKeyGenerationMessage);
+  return getEddsaKeyFromEntropy(signature.slice(2));
+}
+
+/**
+ * Generates the EdDSA private key, following https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5 .
+ *
+ * @param entropy - Random entropy to generate key from as hex string.
+ * @returns The eddsa private key as buffer.
+ */
+export function getEddsaKeyFromEntropy(entropy: string): EddsaPrivateKey {
+  let source = entropy;
+  if (entropy.startsWith('0x')) {
+    source = entropy.slice(2);
+  }
+  if (source.length < 64) {
+    throw new Error('Entropy must be at least 32 bytes long');
+  }
+  if (!isHex(source)) {
+    throw new Error('Entropy must be a hex string');
+  }
+
+  // According to https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5 :
+  // The private key is 32 octets (256 bits) of cryptographically secure random data.
+  // So we can just take the first 32 bytes of the entropy.
+  const privKey = Buffer.from(entropy, 'hex').subarray(0, 32);
```

Comment:
Actually, this appeared to be more complicated problem than I thought, because these two comments are contradicting each other. The beast I came up with is:

```typescript
const buf = Buffer.alloc(32);
const n = buf.write('f', 'hex');
```

Thus, you can throw an error if the string was too short, *or* it was in the wrong encoding — you must check return value of `write` for this. But this will not hang the system if an attacker will provide a ridiculously long string.

Reply:
Sincerely, I'm not sure if we really need to defend from such an attack, or I'm over-engineering. But I'll leave it here anyway.

Reply:
Fixing the suggestions above I found another error here. It should be source, not entropy, otherwise an invalid  '0x' gets read into the Hex buffer.

Reply:
I can not follow why it is more complicated than you thought? After checking that the input is a hex string with sufficient length, we can slice the hex string before turning it into a buffer, right?

Reply:
> After checking that the input is a hex string

The problem is already here :) If a malicious actor could force our regex run on really long string (let's say that consumes 2GB of memory), it's a vulnerability like this http://www.securityspace.com/smysecure/catid.html?id=1.3.6.1.4.1.25623.1.0.11175

Reply:
And if you slice a string with `str.slice(0, 64)`, you slice 64 code units, not 64 bytes.

Reply:
Alright, then I

1. Check that it has at least 64 code units
2. Slice it to 64 code units
3. Check that these 64 code units are all in [0-9a-fA-F] => valid hex string
4. Turn it into a 32 byte buffer

This covers everything, doesn't it?

Reply:
- Code units, such as `'äáöü'` are filtered out in 3.
- The 2GB attack is prevented because we only process the first 64 code units.

Reply:
According to [this](https://unicode.org/faq/utf_bom.html#gen6), 64 code units is 128 bytes maximum (in my tests the actual ratio was x3 if it involved zero-width space, which make it 192 bytes). This is still a constant, which seems to make us safe from this attack vector.

I still like my solution more though :) But now they are almost equivalent, so it's fine.

Reply:
> Code units, such as `'äáöü'`

Just in case: those are code points, not code units. Code unit is a fixed-size block, where size depends on encoding (8 bit in UTF-8, 16 bit in UTF-16). Code point is something that has unicode code in form U+1F600 (smile emoji) or U+0061 (latin small letter a); code points might be encoded with multiple code units. You see [the Unicode glossary](https://www.unicode.org/glossary) for reference.